### PR TITLE
Move non-native Processing providers into independent plugins

### PR DESCRIPTION
**Pull request by alexbruy**
## Description

Following discussion in the [QEP 226](https://github.com/qgis/QGIS-Enhancement-Proposals/issues/226) separate SAGA, GRASS and OTB providers into independent plugins. This should prevent issues with loading of the Processing plugin in cases when one of these providers is broken. Also it will make easier turning them into 3rd party plugins in the future if we will make such decision.

SAGA and GRASS plugins are enabled by default to provide the same UX as before. OTB provider can be enabled in the Plugin Manager if necessary. Enabling/disabling of these three providers now performed via the Plugin Manager like with any other plugin.

This is mostly an internal change, the only visible for users part is using Plugin Manager to activate or deactivate SAGA, GRASS and OTB providers.

### CODE REVIEW

File: python/plugins/sagaprovider/SagaProviderPlugin.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+
+class SagaProviderPlugin:
+
+    def __init__(self):
+        self.provider = SagaAlgorithmProvider()
+
+    def initGui(self):
+        QgsApplication.processingRegistry().addProvider(self.provider)
+
```
Comment (nyalldawson) [inline python/plugins/sagaprovider/SagaProviderPlugin.py:37]: This also should initProcessing so it works nicely with qgis_process

File: python/plugins/sagaprovider/SagaAlgorithmProvider.py [hunk 2]
```diff
@@ -53,8 +53,6 @@ def __init__(self):
     def load(self):
         with QgsRuntimeProfiler.profile('SAGA Provider'):
             ProcessingConfig.settingIcons[self.name()] = self.icon()
-            ProcessingConfig.addSetting(Setting("SAGA", 'ACTIVATE_SAGA',
```
Comment (nyalldawson) [inline python/plugins/sagaprovider/SagaAlgorithmProvider.py:56]: Good move 👍

File: python/plugins/grassprovider/GrassProviderPlugin.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+
+class GrassProviderPlugin:
+
+    def __init__(self):
+        self.provider = Grass7AlgorithmProvider()
+
+    def initGui(self):
+        QgsApplication.processingRegistry().addProvider(self.provider)
+
```
Comment (nyalldawson) [inline python/plugins/grassprovider/GrassProviderPlugin.py:37]: Also needs initProcessing

File: python/plugins/grassprovider/GrassProviderPlugin.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+__author__ = 'Alexander Bruy'
+__date__ = 'June 2021'
+__copyright__ = '(C) 2021, Alexander Bruy'
+
+
+from qgis.core import QgsApplication
+
+from grassprovider.Grass7AlgorithmProvider import Grass7AlgorithmProvider
+
```
Comment (nyalldawson) [inline python/plugins/grassprovider/GrassProviderPlugin.py:28]: Could you re-add the runtime profiler block here?

File: python/plugins/otbprovider/OtbProviderPlugin.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+
+class OtbProviderPlugin:
+
+    def __init__(self):
+        self.provider = OtbAlgorithmProvider()
+
+    def initGui(self):
+        QgsApplication.processingRegistry().addProvider(self.provider)
+
```
Comment (nyalldawson) [inline python/plugins/otbprovider/OtbProviderPlugin.py:37]: initProcessing

File: python/plugins/otbprovider/OtbProviderPlugin.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+"""
+
+__author__ = 'Alexander Bruy'
+__date__ = 'June 2021'
+__copyright__ = '(C) 2021, Alexander Bruy'
+
+
+from qgis.core import QgsApplication
+
+from otbprovider.OtbAlgorithmProvider import OtbAlgorithmProvider
```
Comment (nyalldawson) [inline python/plugins/otbprovider/OtbProviderPlugin.py:27]: Re-add runtime profiler block here

**Review by nyalldawson — APPROVED** [review]
Great -- this is a fantastic step forward!

**Comment by nyalldawson** [comment]
> probably it is better to copy all necessary test data and internal test code (e.g. AlgorithmsTestBase.py) to each plugin.

I'm +1 to this approach

**Comment by nyalldawson** [comment]
Hey @alexbruy what do you think should be done regarding the gdal algorithms? I'm 50/50 on whether they should also be split out. I never want to see them dropped from the master QGIS package, but from a purely "keep everything partitioned" point of view it would be cleaner to also separate these out. I'm keen to hear your thoughts though!

(we could even consider moving the remaining QGIS python ones out to a separate plugin if we're going down this route, so that the Processing plugin is ONLY the framework components (i.e. the bits which will ultimately be ported to c++)

File: python/plugins/grassprovider/GrassProviderPlugin.py (hunk 3 above)
Comment (alexbruy) [reply python/plugins/grassprovider/GrassProviderPlugin.py:38]: Done.

File: python/plugins/otbprovider/OtbProviderPlugin.py (hunk 5 above)
Comment (alexbruy) [reply python/plugins/otbprovider/OtbProviderPlugin.py:38]: Done.

File: python/plugins/otbprovider/OtbProviderPlugin.py (hunk 6 above)
Comment (alexbruy) [reply python/plugins/otbprovider/OtbProviderPlugin.py]: Added.

**Comment by alexbruy** [comment]
@nyalldawson I don't thing we should drop GDAL algorithms from the core. We can separate it into a plugin to keep things organized and clean. Or maybe even port them to C++ and move to the analysis library as we do with native algorithms. As QGIS can not be built without GDAL at least part of the GDAL algorithms (these implemented as cli apps) always will be available.

But I don't have a strong opinion here as well.

**Comment by nyalldawson** [comment]
@alexbruy 
> I don't thing we should drop GDAL algorithms from the core. 

I totally agree -- what I'm wondering is if we should split it into a separate "core plugin" still, in order to keep the main processing plugin as light as possible.

**Comment by alexbruy** [comment]
@nyalldawson I think, from the maintenance point of view it makes sense. Do you think it is worth to move them to the analysis library or is enough to only separate?

File: python/plugins/grassprovider/GrassProviderPlugin.py (hunk 4 above)
Comment (alexbruy) [reply python/plugins/grassprovider/GrassProviderPlugin.py:29]: Added.

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
@alexbruy 
> Do you think it is worth to move them to the analysis library or is enough to only separate?

I honestly can't see many benefits in moving them to c++ ... the GDAL ones have been so pain free to maintain anyway

**Comment by MrChebur** [comment]
Has this change affected the way GRASS algorithms are imported into the Standalone scripts? I would be grateful for an answer here: https://gis.stackexchange.com/questions/420721/running-the-grass-algorithm-in-standalone-python-code-using-the-pyqgis-processin

**Comment by agiudiceandrea** [comment]
Hi @MrChebur, is your issue the same described in the https://github.com/qgis/QGIS/issues/45935 issue report?

**Pull request merged**